Platform connections: Add new data platform adapter for CrateDB

amotl opened the pull request:
## About
First things first, thanks a stack for conceiving and maintaining dbt. 💯

This patch adds information about the database adapter for the [CrateDB database](https://github.com/crate/crate) to dbt's documentation. Because CrateDB is compatible with PostgreSQL, it builds heavily upon the [dbt-postgres](https://github.com/dbt-labs/dbt-postgres) adapter.

It was tremendously helpful for us to get started by gradually working through its test suite to progressively make test cases succeed on CrateDB, so we would like to explicitly send kudos to @mikealfare and all contributors.

## Preview
[CrateDB setup](https://docs-getdbt-com-git-fork-crate-workbench-cratedb-dbt-labs.vercel.app/docs/core/connect-data-platform/cratedb-setup)

## Details
**Package:** https://pypi.org/project/dbt-cratedb2/
**Repository:** https://github.com/crate/dbt-cratedb2
**Documentation:** https://cratedb.com/docs/guide/integrate/dbt/

## Checklist
ADDING OR REMOVING PAGES:
- [x] Add/remove page in `website/sidebars.js`
- [x] Provide a unique filename for new pages
- [ ] Add an entry for deleted pages in `website/vercel.json`
- [ ] Run link testing locally with `npm run build` to update the links that point to deleted pages

amotl committed:
Resource configurations: Improve `no-configs` page

- Fix a typo s/specifc/specific/
- Add third variant why no documentation about platform-specific
  configuration settings is provided, by referring to vendor
  documentation pages

amotl commented:
Dear @mirnawong1,

I just wanted to check if you need anything from my end to complete this patch, for example by improving it?

With kind regards,
Andreas.

mirnawong1 commented on website/docs/docs/core/connect-data-platform/cratedb-setup.md:
```diff
@@ -0,0 +1,61 @@
+dbt with CrateDB], which also includes up-to-date information.
+
+
+## Profile Configuration
```
```suggestion
## Profile configuration
```

amotl replied:
Fixed. Thanks.

amotl replied:
Fixed, thanks.

mirnawong1 commented on website/docs/docs/core/connect-data-platform/cratedb-setup.md:
```diff
@@ -0,0 +1,61 @@
+  github_repo: 'crate/dbt-cratedb2'
+  pypi_package: 'dbt-cratedb2'
+  min_core_version: 'v1.0.0'
+  cloud_support: Supported
```
hey @amotl  just double checking if cratedb is dbt cloud-supported?

amotl replied:
That probably was wishful thinking. Thanks for spotting, fixed.

mirnawong1 commented:
looping in @amychen1776 for review

amychen1776 commented:
I'm going to enable the testing so I can get a preview of this PR

amychen1776 commented on website/docs/docs/core/connect-data-platform/cratedb-setup.md:
```diff
@@ -0,0 +1,61 @@
+      pass: [password]
+      dbname: crate     # CrateDB's only catalog is `crate`.
+      schema: doc       # Define schema. `doc` is the default.
+      search_path: doc  # Use the same value like `schema` by default.
```
What is the difference between schema and search path (or maybe the better question is how are they related?)

amotl replied:
We have a bit of information about `search_path` on the documentation page about [CrateDB session settings](https://cratedb.com/docs/crate/reference/en/latest/config/session.html). In general, it is the same like the [PostgreSQL setup](https://docs.getdbt.com/docs/core/connect-data-platform/postgres-setup) page says:

> [search_path](https://docs.getdbt.com/docs/core/connect-data-platform/postgres-setup#search_path): [optional, override the default postgres `search_path`]

Maybe we should align with the PostgreSQL page, but explicitly expand into stating the exact differences to PostgreSQL:

> The `search_path` config controls the Postgres "search path" that dbt configures when opening new connections to the database. By default, the Postgres search path is ~~"$user, public"~~ "pg_catalog, doc", meaning that unqualified table names will be searched for in the `doc` schema, ~~or a schema with the same name as the logged-in user.~~ **Note:** Setting the `search_path` to a custom value is not necessary or recommended for typical usage of dbt.

In this case, it might be advisable to remove this setting completely from the minimal set of settings we are presenting on the upcoming [CrateDB setup page](https://docs-getdbt-com-git-fork-crate-workbench-cratedb-dbt-labs.vercel.app/docs/core/connect-data-platform/cratedb-setup), and only describe it on the [CrateDB and dbt documentation page](https://cratedb.com/docs/guide/integrate/dbt/) instead. wdyt? /cc @kneth

Can you please also review this, @seut and @hlcianfagna, to confirm that this also holds true for CrateDB:

> **Note:** Setting the `search_path` to a custom value is not necessary or recommended for typical usage of dbt.

amotl replied:
Hi again. I've just updated the patch, removing the option about `search_path`. It also works without this setting. We decided to keep the basic configuration section minimal, and the CrateDB setup page is referring to the PostgreSQL setup page anyway.

Please let us know if you think differently, @kneth, @hlcianfagna, or @seut, possibly also in retrospective.

hlcianfagna replied:
It is an advanced option, I agree it may be better omitted in the minimal example as it is optional and may lead to confusion.

amotl replied:
Thank you. Resolving this.

amychen1776 commented on website/docs/docs/core/connect-data-platform/cratedb-setup.md:
```diff
@@ -0,0 +1,61 @@
+[CrateDB]: https://cratedb.com/database
+[Postgres profile setup]: /docs/core/connect-data-platform/postgres-setup
+[`profiles.yml`]: https://docs.getdbt.com/docs/core/connect-data-platform/profiles.yml
+[using dbt with CrateDB]: https://cratedb.com/docs/guide/integrate/dbt/
```
as a heads up for your own documentation - we no longer called dbt data build tool. It's just `dbt`  and is no longer broken out

amotl replied:
Thank you. This patch fixes it.

- https://github.com/crate/cratedb-guide/pull/161.

amychen1776 reviewed (commented):
Are there any configurations that need to be mentioned in terms of using the adapter like we do for postgres? https://docs.getdbt.com/reference/resource-configs/postgres-configs

amotl commented on website/docs/docs/core/connect-data-platform/cratedb-setup.md:
```diff
@@ -0,0 +1,61 @@
+  slack_channel_name: 'Community Forum'
+  slack_channel_link: 'https://community.cratedb.com/'
+  platform_name: 'CrateDB'
+  config_page: '/reference/resource-configs/postgres-configs'
```
_Let me reiterate this here, to have a threaded conversation about this topic._

@amychen1776 asked:

> Are there any configurations that need to be mentioned in terms of using the adapter like we do for postgres? https://docs.getdbt.com/reference/resource-configs/postgres-configs

amotl replied:
This is right on the spot, @amychen1776. Thanks for bringing it up.

I've sloppily referred to the PostgreSQL configuration settings here, because I thought to have picked up this strategy from another page that describes a database that is compatible with PostgreSQL. However, I might be wrong, as I can't find it back, and it is a bad idea anyway.

Actually, we need to verify first what CrateDB will be capable of, because we needed to deactivate a few test cases compared to PostgreSQL, where CrateDB lacks relevant features.

- https://github.com/crate/dbt-cratedb2/pull/2

amotl replied:
In this spirit, we should review our implementation, and then possibly add a dedicated cratedb-configs page, for example like [yellowbrick-configs](https://docs.getdbt.com/reference/resource-configs/yellowbrick-configs) is doing it, that educates readers about the fact CrateDB might only support a subset of the materialization strategies compared to what the PostgreSQL adapter provides per [postgres-configs](https://docs.getdbt.com/reference/resource-configs/postgres-configs).

Another option, possibly for immediate application, would be to do it like the [risingwave-setup](https://docs.getdbt.com/docs/core/connect-data-platform/risingwave-setup) page is doing it, that is referring to the [no-configs](https://docs.getdbt.com/reference/resource-configs/no-configs) page, that says there are  "no specific configurations for this adapter". This way, we could contribute a relevant page later, but sure enough it would be sweet to get it right and complete from the very beginning.

Can you possibly support here, @hlcianfagna?

amotl replied:
For pragmatic reasons, mostly to unlock the patch to be merged, I've adjusted it to refer to the [no-configs](https://docs.getdbt.com/reference/resource-configs/no-configs) page now, and created a downstream ticket to track that gap on our realms.

- https://github.com/crate/dbt-cratedb2/issues/7

067d8b819e, included here, slightly improves the no-configs page on your realms. If you don't like it, let me know if I should remove it again. Thanks!

hlcianfagna replied:
I second the idea of going with the no-configs page for the time being. I added a couple of thoughts to [the follow up ticket](https://github.com/crate/dbt-cratedb2/issues/7).

amotl replied:
Thank you. Resolving this.

amychen1776 replied:
Oh somehow I couldn't find this comment thread until now but I recommend adding your own dedicated config ticket in the future but will resolve my comment for now so we can merge this PR in.

amotl replied:
Thanks!

amotl reviewed (commented):
Thank you for the review, @amychen1776. I see we need to clarify two of the details you flagged, so I've tagged a few of my colleagues, to kindly ask them for feedback.

amotl committed:
Platform connections: Add new data platform adapter for CrateDB

The Python package is called `dbt-cratedb2`.

Package: https://pypi.org/project/dbt-cratedb2/
Repository: https://github.com/crate/dbt-cratedb2
Documentation: https://cratedb.com/docs/guide/integrate/dbt/

amotl commented:
Hi again. I think your suggestions have been resolved. Thank you!

amychen1776 reviewed (commented):
Checking the preview - you're missing the link in the Community Adapter section https://docs-getdbt-op402z6fm-dbt-labs.vercel.app/docs/community-adapters (unless you want to be an trusted adapter, then you would want to be linked in the trusted adapter section)

amychen1776 commented on website/docs/docs/core/connect-data-platform/cratedb-setup.md:
```diff
@@ -0,0 +1,62 @@
+
+CrateDB targets are configured exactly the same way, see also [PostgreSQL
+configuration], with just a few things to consider which are special to
+CrateDB. Relevant details are outlined at [using dbt with CrateDB],
```
This is a bit confusing because this link actually goes to no information. If there are no special configurations, I would just not even add this sentence for a better user experience.

amychen1776 replied:
@mirnawong1 would welcome your thoughts here

amychen1776 replied:
Found thread and comments. Will resolve now.

amotl commented on website/docs/docs/community-adapters.md:
```diff
@@ -7,7 +7,8 @@ Community adapters are adapter plugins contributed and maintained by members of
 | Data platforms (click to view setup guide) |||
 | ------------------------------------------ | -------------------------------- | ------------------------------------- |
-| [Clickhouse](/docs/core/connect-data-platform/clickhouse-setup) | [Databend Cloud](/docs/core/connect-data-platform/databend-setup) | [Doris & SelectDB](/docs/core/connect-data-platform/doris-setup)  |
+| [Clickhouse](/docs/core/connect-data-platform/clickhouse-setup) | [CrateDB](/docs/core/connect-data-platform/cratedb-setup)
+| [Databend Cloud](/docs/core/connect-data-platform/databend-setup) | [Doris & SelectDB](/docs/core/connect-data-platform/doris-setup)  |
```
> you're missing the link in the Community Adapter section

Thank you. Fixed.

amotl replied:
> unless you want to be an trusted adapter, then you would want to be linked in the trusted adapter section

I guess that would not be reasonable for a 0.0.2 version. Maybe @kneth would like to follow up with you on that later. What would it take to become a trusted adapter?

amotl replied:
Right, it's [on the docs](https://docs-getdbt-com-git-fork-crate-workbench-cratedb-dbt-labs.vercel.app/docs/trusted-adapters). This sounds very good.

amotl replied:
I guess [What it means to be trusted](https://docs-getdbt-com-git-fork-crate-workbench-cratedb-dbt-labs.vercel.app/guides/adapter-creation#what-it-means-to-be-trusted) includes all the important details?
I've created a ticket about it, to do it later, thanks!

- https://github.com/crate/dbt-cratedb2/issues/9

amychen1776 replied:
That's correct!

amychen1776 reviewed (approved):
:shipit:

amotl commented:
Thank you very much for the excellent guidance. 🌻

mirnawong1 commented on website/docs/docs/core/connect-data-platform/cratedb-setup.md:
```diff
@@ -0,0 +1,62 @@
+
+[CrateDB]: https://cratedb.com/database
+[PostgreSQL configuration]: https://docs.getdbt.com/reference/resource-configs/postgres-configs
+[PostgreSQL profile setup]: /docs/core/connect-data-platform/postgres-setup
```
```suggestion
[PostgreSQL profile setup]( /docs/core/connect-data-platform/postgres-setup)
```

amotl replied:
Oh.

amotl replied:
Fixed all links, thanks. I am using Sphinx/MyST, that's why I was used to the other out-of-band link syntax. Sorry that I did not look deeply enough how to run the linter on my workstation.

mirnawong1 commented on website/docs/docs/core/connect-data-platform/cratedb-setup.md:
```diff
@@ -0,0 +1,62 @@
+
+
+
+[CrateDB]: https://cratedb.com/database
```
```suggestion
[CrateDB](https://cratedb.com/database)
```

mirnawong1 replied:
not sure if you meant to have the text hyperlinked so i removed the colon so it rendered appropriately

amotl replied:
Yes, all those were supposed to be Hyperlinks. However, apparently, they did not render properly on the [preview](https://docs-getdbt-com-git-fork-crate-workbench-cratedb-dbt-labs.vercel.app/docs/core/connect-data-platform/cratedb-setup) after applying your suggestion, see screenshots below.

So, I reverted those changes again. Please advise.

mirnawong1 commented on website/docs/docs/core/connect-data-platform/cratedb-setup.md:
```diff
@@ -0,0 +1,62 @@
+
+
+[CrateDB]: https://cratedb.com/database
+[PostgreSQL configuration]: https://docs.getdbt.com/reference/resource-configs/postgres-configs
```
```suggestion
[PostgreSQL configuration](/reference/resource-configs/postgres-configs)
```

mirnawong1 commented on website/docs/docs/core/connect-data-platform/cratedb-setup.md:
```diff
@@ -0,0 +1,62 @@
+[CrateDB]: https://cratedb.com/database
+[PostgreSQL configuration]: https://docs.getdbt.com/reference/resource-configs/postgres-configs
+[PostgreSQL profile setup]: /docs/core/connect-data-platform/postgres-setup
+[`profiles.yml`]: https://docs.getdbt.com/docs/core/connect-data-platform/profiles.yml
```
```suggestion
[`profiles.yml`](/docs/core/connect-data-platform/profiles.yml)
```

mirnawong1 commented on website/docs/docs/core/connect-data-platform/cratedb-setup.md:
```diff
@@ -0,0 +1,62 @@
+[PostgreSQL configuration]: https://docs.getdbt.com/reference/resource-configs/postgres-configs
+[PostgreSQL profile setup]: /docs/core/connect-data-platform/postgres-setup
+[`profiles.yml`]: https://docs.getdbt.com/docs/core/connect-data-platform/profiles.yml
+[using dbt with CrateDB]: https://cratedb.com/docs/guide/integrate/dbt/
```
```suggestion
[using dbt with CrateDB](https://cratedb.com/docs/guide/integrate/dbt/)
```

mirnawong1 commented:
great stuff, thank you @amotl and @amychen1776 ! happy to merge this ?

amotl commented on website/docs/docs/core/connect-data-platform/cratedb-setup.md:
```diff
@@ -0,0 +1,62 @@
+
+## Profile configuration
+
+CrateDB targets should be set up using a configuration like this minimal sample
+of settings in your [`profiles.yml`] file.
```
There is currently a hiccup.

![image](https://github.com/user-attachments/assets/59277975-0336-47d4-b1c7-75f3e778af39)

amotl replied:
Hm. I think all the links worked well before.

![image](https://github.com/user-attachments/assets/cb57b093-f531-41b3-8967-46b80b6e52c4)

Maybe I did a mistake, misunderstanding your advises.

amotl replied:
Let me revert the changes and then let us have a look at the preview rendering again.

amotl replied:
I think all links are rendering well now?

-- https://docs-getdbt-com-git-fork-crate-workbench-cratedb-dbt-labs.vercel.app/docs/core/connect-data-platform/cratedb-setup

mirnawong1 replied:
they do work! thanks @amotl !

mirnawong1 commented:
hey @amotl , have enabled auto merge. any chance you can update this pr with the base branch?

amotl commented:
@mirnawong1: Thanks. I think the job still needs attention once more. Because I am an outsider, auto-merge was disabled again.

> auto-merge was automatically disabled [December 11, 2024 18:11](https://github.com/dbt-labs/docs.getdbt.com/pull/6564#event-15621999967). Head branch was pushed to by a user without write access.
